pkg/storage/trb_request_document.go

```diff
@@ -0,0 +1,168 @@
+
+import (
+	"fmt"
+
+	"github.com/google/uuid"
+	"go.uber.org/zap"
+
+	"github.com/cmsgov/easi-app/pkg/apperrors"
+	"github.com/cmsgov/easi-app/pkg/models"
+)
+
+// GetTRBRequestDocumentsByRequestID queries the DB for all documents attached to the TRB request with the given ID
+func (s *Store) GetTRBRequestDocumentsByRequestID(logger *zap.Logger, requestID uuid.UUID) ([]*models.TRBRequestDocument, error) {
+	const trbRequestDocumentsGetByRequestIDSQL = `
+		SELECT id,
+			trb_request_id,
+			file_name,
+			document_type,
+			other_type,
+			bucket,
+			s3_key,
+			created_by,
+			modified_by,
+			created_at,
+			modified_at
+		FROM trb_request_documents
+		WHERE trb_request_id = :trb_request_id
+	`
+
+	documents := []*models.TRBRequestDocument{}
+
+	stmt, err := s.db.PrepareNamed(trbRequestDocumentsGetByRequestIDSQL)
+	if err != nil {
+		return nil, err
+	}
+
+	arg := map[string]interface{}{
+		"trb_request_id": requestID,
+	}
+
+	err = stmt.Select(&documents, arg)
+	if err != nil {
+		logger.Error(
+			"Failed to fetch TRB request documents for request ID "+requestID.String(),
+			zap.Error(err),
+			zap.String("requestID", requestID.String()),
+		)
+		return nil, &apperrors.QueryError{
+			Err:       err,
+			Model:     models.TRBRequestDocument{},
+			Operation: apperrors.QueryFetch,
+		}
+	}
+
+	return documents, nil
+}
+
+// CreateTRBRequestDocument creates a record for a TRBRequestDocument in our database, *after* it's been uploaded to S3
+func (s *Store) CreateTRBRequestDocument(logger *zap.Logger, document *models.TRBRequestDocument) (*models.TRBRequestDocument, error) {
+	const trbRequestDocumentCreateSQL = `
+		INSERT INTO trb_request_documents (
+			id,
+			trb_request_id,
+			file_name,
+			document_type,
+			other_type,
+			bucket,
+			s3_key,
+			created_by,
+			modified_by
+		) VALUES (
+			:id,
+			:trb_request_id,
+			:file_name,
+			:document_type,
+			:other_type,
+			:bucket,
+			:s3_key,
+			:created_by,
+			:modified_by
+		) RETURNING
+			id,
+			trb_request_id,
+			file_name,
+			document_type,
+			other_type,
+			bucket,
+			s3_key,
+			created_by,
+			created_at,
+			modified_by,
+			modified_at
+	`
+
+	if document.ID == uuid.Nil {
+		document.ID = uuid.New()
+	}
+
+	stmt, err := s.db.PrepareNamed(trbRequestDocumentCreateSQL)
+	if err != nil {
+		return nil, err
+	}
+
+	retDoc := models.TRBRequestDocument{}
+	err = stmt.Get(&retDoc, document)
+	if err != nil {
+		logger.Error(
+			fmt.Sprintf("Failed to create trb request document with error %s", err),
+			zap.Error(err),
+			zap.String("user", document.CreatedBy),
+		)
+		return nil, &apperrors.QueryError{
+			Err:       err,
+			Model:     document,
+			Operation: apperrors.QueryPost,
+		}
+	}
+
+	return &retDoc, nil
+}
+
+// DeleteTRBRequestDocument deletes an existing TRBRequestDocument, given its ID
+func (s *Store) DeleteTRBRequestDocument(logger *zap.Logger, id uuid.UUID) (*models.TRBRequestDocument, error) {
+	const trbRequestDocumentDeleteSQL = `
+		DELETE
```

Comment:
I like this suggestion. Though, in the interest of getting this work through and unblocked, it might be better captured as an improvement ticket to come back to (especially since these TRB features are behind a flag and not available to users yet).

@jstrang-oddball would you be able to create a ticket to capture the soft-delete functionality? I'm down to hop on a call and talk about it too, because I have some other ideas about how we could (at a greater scale) improve our document upload/storage process, and maybe they could go in a ticket together!

Reply:
I'm definitely in favor of creating a new ticket for improving the deletion process, so we can get this ticket's work done. I created [EASI-2453](https://jiraent.cms.gov/browse/EASI-2453) and pinged Bill, Farrah, and Jeff on it for prioritization.

I don't think we need to do soft deletion as part of this ticket, since it'll still be a while before the TRB workflow's in use; if we really need to add soft-deletion to enable further work, we can implement it before the TRB goes live.